SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
+if(density_occuiped_ == nulltype_()){
+throw std::invalid_argument("density_occupied has not been registered. please assign the"
+"space label by calling assign_density_occupied(label)");
+}
+else{
+return density_occuiped_;
+}
+}
+void assign_vacuum_occupied(std::wstring label){
+  if(label_space.find(label) == label_space.end()){
+  throw std::invalid_argument("label not added to registry");
+}
+vacuum_occupied_ = label_space[label];
+}
+
+void assign_complete(std::wstring label){
+if(label_space.find(label) == label_space.end()){
+  throw std::invalid_argument("label not added to registry");
+}
+complete_ = label_space[label];
+}
+
+void assign_active_particle_space(std::wstring label){
+  if(label_space.find(label) == label_space.end()){
+  throw std::invalid_argument("label not added to registry");
+}
+active_particle_space_ = label_space[label];
+}
+
+void assign_active_hole_space(std::wstring label){
+  if(label_space.find(label) == label_space.end()){
+  throw std::invalid_argument("label not added to registry");
+}
+active_hole_space_ = label_space[label];
+}
+
+// which spaces could contain particles in your representation
+void assign_density_occupied(std::wstring label){
+if(label_space.find(label) == label_space.end()){
+throw std::invalid_argument("label not added to registry");
+}
+density_occuiped_ = label_space[label];
+}
+
+
+ private:
+  std::map<std::wstring ,IndexSpace> label_space;
+  const IndexSpace nulltype= {L"",0,0};
+
+  ///TODO use c++20 std::has_single_bit() when we update to this version
+  bool has_single_bit(std::uint32_t bits){
+    return bits&(((bool)(bits&(bits-1)))-1);
+  }
+  // find an indexspace from its type. return nullspace if not present.
+  // a bit strange, but prevents an additional map that needs to be maintained
+  const IndexSpace find_IndexSpace(IndexSpace IS)const{
+    for(auto it = label_space.begin(); it != label_space.end(); it++){
+      if(it->second.attr() == IS.get_attr()){
+        return it->second;
+      }
+    }
+    return nulltype;
   }
 
-  static bool instance_exists(Attr attr) {
-    return instances_.find(attr) != instances_.end();
+  // find an IndexSpace from attribute. return nullspace if not present.
+  // sometimes we wish to check whether if an antribute is in the registry
+const IndexSpace find_IndexSpace(IndexSpace::Attr attr)const{
+  for(auto it = label_space.begin(); it != label_space.end(); it++){
+    if(it->second.attr() == attr){
+      return it->second;
+    }
   }
+  return nulltype;
+}
+  IndexSpace complete_ = {L"",0,0}; // need to define to use generic operator class
+  IndexSpace vacuum_occupied_ = {L"",0,0}; // needed for fermi vacuum wick application
+  IndexSpace density_occuiped_ = {L"",0,0};
+  // both needed to make excitation and de-excitation operators. not neccessarily exclusionary in the case of
+  // multi-reference context.
+  IndexSpace active_particle_space_ = {L"",0,0};
+  IndexSpace active_hole_space_ = {L"",0,0};
```

Comment:
Why not use default constructors?
```suggestion
  IndexSpace complete_; // need to define to use generic operator class
  IndexSpace vacuum_occupied_; // needed for fermi vacuum wick application
  IndexSpace density_occupied_;
  // both needed to make excitation and de-excitation operators. not neccessarily exclusionary in the case of
  // multi-reference context.
  IndexSpace active_particle_space_;
  IndexSpace active_hole_space_;
```